### PR TITLE
feat: add git-audit support when building the package artifacts

### PR DESCRIPTION
**Pull request by jenstroeger**
Related to https://github.com/jenstroeger/python-package-template/issues/5#issuecomment-1207180258: runs [`git-audit`](https://github.com/pypa/pip-audit) when the `requirements.txt` file is generated.

@behnazh this PR has two commits where I tried to use `git-audit` in two different ways:
- commit https://github.com/jenstroeger/python-package-template/pull/307/commits/19cbd9ec18d0cd1ecc7a951fab5a8c7be4dae5c9: as part of the local Makefile when generating the `requirements.txt` file — but that adds a new dependency to the `pip-audit` package for a _pruned_ tree which makes the [run](https://github.com/jenstroeger/python-package-template/runs/8180973112?check_suite_focus=true#step:9:139) a little ugly and adds a hard-wired dependency into our Makefile;
- commit https://github.com/jenstroeger/python-package-template/pull/307/commits/aae32eddb2f4d38ba79c4f1fcccf4dbd154fbc0a: as a Github Action — which required a little massaging of the Makefile when it builds the `requirements.txt` because we can’t `git-audit` the package itself and had to exclude it from the `requirements.txt` for the audit [run](https://github.com/jenstroeger/python-package-template/runs/8181083972?check_suite_focus=true#step:10:1).<sup>*</sup>

I wonder if we should document that `git-audit` runs and checks all dependent packages?

—————
<sup>*</sup> And now the Action installs the `pip-audit` package…

### CODE REVIEW

**Comment by jenstroeger** [comment]
As a third alternative (as discussed) here is the change that audits _all_ packages instead of only shipping packages:
```diff
index b7c0b3c..302cfad 100644
--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -60,19 +60,25 @@ jobs:
       # those targets first and only builds the package if they succeed.
       # Build the sdist and wheel distribution of the package and docs as a zip file.
       run: make dist
+    # Generate the requirements.txt that contains the hash digests of the dependencies and
+    # generate the SBOM using CyclonDX SBOM generator.
+    - name: Generate requirements and SBOM for all packages
+      run: make requirements sbom
+    # Audit all required packages for security vulnerabilities.
+    - name: Audit required packages
+      uses: pypa/gh-action-pip-audit@v1.0.0
+      with:
+        inputs: requirements.txt
+    # Remove the old requirements.txt file (which includes _all_ packages) and generate a
+    # new one for the package and its actual and required dependencies only.
+    - name: Pruning requirements for shipping package
+      run: make prune requirements
     - name: Compute package hash
       if: matrix.os == env.ARTIFACT_OS && matrix.python == env.ARTIFACT_PYTHON
       id: compute-hash
       shell: bash
       run: |
         set -euo pipefail
-        # Generate the requirements.txt that contains the hash digests of the dependencies and
-        # generate the SBOM using CyclonDX SBOM generator.
-        make requirements sbom
-        # Remove the old requirements.txt file (which includes _all_ packages) and generate a
-        # new one for the package and its actual and required dependencies only.
-        rm requirements.txt
-        make prune requirements
         # Find the paths to the files that will be included in the release.
         TARBALL_PATH=$(find dist -name "*.tar.gz")
         WHEEL_PATH=$(find dist -name "*.whl")
@@ -85,12 +91,6 @@ jobs:
         echo "Digest of artifacts is $DIGEST."
         # Set the computed sha digest as the output of this job.
         echo "::set-output name=artifacts-sha256::$DIGEST"
-    # Audit the required packages for security vulnerabilities.
-    - name: Audit required packages
-      if: matrix.os == env.ARTIFACT_OS && matrix.python == env.ARTIFACT_PYTHON
-      uses: pypa/gh-action-pip-audit@v1.0.0
-      with:
-        inputs: requirements.txt
     # For now only generate artifacts for the specified OS and Python version in env variables.
     # Currently reusable workflows do not support setting strategy property from the caller workflow.
     - name: Upload the package artifact for debugging and release
```
However, I think this is a disagreeable approach because the [pip-audit Action](https://github.com/pypa/gh-action-pip-audit) installs a whole bunch of additional packages (see [this run](https://github.com/jenstroeger/python-package-template/runs/8181427245?check_suite_focus=true#step:10:36)) which introduces new packages and dependencies _outside_ of the SBOM we just created.

**Comment by jenstroeger** [comment]
Commit https://github.com/jenstroeger/python-package-template/pull/307/commits/2ee985a2051cba05c563ec38c0818ae7c5ff710f adds `pip-audit` as part of the `dev` dependencies, and the Build workflow then audits _all_ packages for _all_ platform runners (see for example [this run](https://github.com/jenstroeger/python-package-template/runs/8189210153?check_suite_focus=true#step:10:1)). I also added a new `audit` goal to the Makefile such that users can audit their packages locally.

Should we mention the use of `bandit` and `pip-audit` in the [Security Analysis](https://github.com/jenstroeger/python-package-template#security-analysis) section?

**Comment by jenstroeger** [comment]
See issue https://github.com/pypa/pip-audit/issues/365 for the failing runs on Windows.

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -60,19 +60,25 @@ jobs:
...
+      run: make requirements sbom
+    # Audit all required and currently installed packages for security vulnerabilities.
+    - name: Audit required packages
+      run: make audit
+    # Remove the old requirements.txt file (which includes _all_ packages) and generate a
+    # new one for the package and its actual and required dependencies only.
+    - name: Prune requirements for shipping package
+      run: |
+        rm requirements.txt
+        make prune requirements
```
Comment (jenstroeger) [inline .github/workflows/build.yaml:75]: Should we add a
```yaml
if: matrix.os == env.ARTIFACT_OS && matrix.python == env.ARTIFACT_PYTHON
```
to these three steps?

Comment (jenstroeger) [reply .github/workflows/build.yaml]: As discussed, let’s run audit before the SBOM: commit https://github.com/jenstroeger/python-package-template/pull/307/commits/5b721256bdf5f7531af4f88fd488765afa5757d7. Keeps this change clean and simple 👍🏼

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -60,6 +60,10 @@ jobs:
       # those targets first and only builds the package if they succeed.
       # Build the sdist and wheel distribution of the package and docs as a zip file.
       run: make dist
+    # Audit all currently installed packages for security vulnerabilities.
+    - name: Audit installed packages
+      if: matrix.os == env.ARTIFACT_OS && matrix.python == env.ARTIFACT_PYTHON
```
Comment (behnazh) [inline .github/workflows/build.yaml:65]: Why not to always run `make audit`?

File: Makefile [hunk 3]
```diff
@@ -130,6 +131,14 @@ requirements.txt: pyproject.toml
 	echo "" >> requirements.txt
 	cp requirements.txt dist/package-$(PACKAGE_VERSION)-requirements.txt
 
+# Audit the currently installed packages.
+.PHONY: audit
+audit:
+	if ! $$(python -c "import pip_audit" &> /dev/null); then \
+	  echo "No package pip_audit installed, upgrade your environment!" && exit 1; \
+	fi;
+	python -m pip_audit --skip-editable --desc on --fix --dry-run
```
Comment (behnazh) [inline Makefile:140]: Can you please add comments for these cmd line options?

File: README.md [hunk 4]
```diff
@@ -59,6 +59,8 @@ Automatic package versioning and tagging, publishing to [PyPI](https://pypi.org/
 
 [OSSF Security Scorecards](https://github.com/ossf/scorecard) is enabled as a GitHub Actions workflow to give the consumers information about the supply-chain security posture of this project, assigning a score of 0–10. We upload the results as a SARIF (Static Analysis Results Interchange Format) artifact after each run and the results can be found at the Security tab of this GitHub project. We also allow publishing the data at [OpenSSF](https://metrics.openssf.org/). We use this data to continuously improve the security posture of this project. Note that this configuration supports the ``main`` (default) branch and requires the repository to be public and not forked.
 
+[pip-audit](https://github.com/pypa/pip-audit) is part of the default Python virtual environment, and can be used to check all installed packages for documented [VCE](https://www.cve.org/) by querying the [Python Packaging Advisory Database](https://github.com/pypa/advisory-database). The `build.yaml` workflow always runs a package audit before the artifacts are being built.
```
Comment (behnazh) [inline README.md:62]: the CVE?

File: .github/workflows/build.yaml (hunk 2 above)
Comment (jenstroeger) [reply .github/workflows/build.yaml]: Sure, why not: commit https://github.com/jenstroeger/python-package-template/pull/307/commits/fb448419629edeb0e5ea152e3110a1c7b8e44af3.

File: Makefile (hunk 3 above)
Comment (jenstroeger) [reply Makefile:143]: Commit https://github.com/jenstroeger/python-package-template/pull/307/commits/12b27833f1328491d2e03023c0281bf667ca8864.

File: README.md (hunk 4 above)
Comment (jenstroeger) [reply README.md]: Nice catch! Commit https://github.com/jenstroeger/python-package-template/pull/307/commits/ff45f27e41d668ea7d77e2c49d5ad9e1d0d3f9bb.

**Pull request merged**